vac/raw/decentralized-messaging-ethereum.md

```diff
@@ -0,0 +1,737 @@
+pk = (p, g, g^a)
+sk = a
+return (pk, sk)
+```
+
+#### PKE-Enc
+
+A user `v` encrypting a message `m` for `u` MUST follow these steps:
+
+```
+PKE-Enc(pk):
+k = randint(1, p-2)
+eta = g^k % p
+delta = m * (g^a)^k % p
+return ((eta, delta))
+```
+
+#### PKE-Dec
+
+The user `u` recovers a message `m` from a ciphertext `c` by performing the following operations:
+
+```
+PKE-Dec(sk):
+mu = eta^(p-1-sk) % p
+return ((mu` * delta) % p)
+```
+
+### DCGKA Syntax
+#### Auxiliary functions
+
+There exist 6 functions that are auxiliary for the rest of components of the protocol, namely:
+
+##### init
+
+This function takes an `ID` as input and returns its associated initial state, denoted by `gamma`:
+
+```
+gamma.myId = ID
+gamma.mySeq = 0
+gamma.history = empty
+gamma.nextSeed = empty_string
+gamma.2sm[·] = empty_string
+gamma.memberSecret[·, ·, ·] = empty_string
+gamma.ratchet[·] = empty_string
+return (gamma)
+```
+
+#### encrypt-to
+
+Upon reception of the recipient’s `ID` and a plaintext, 
+it encrypts a direct message for another group member. 
+Should it be the first message for a particular `ID`, 
+then the `2SM` protocol state is initialized and stored in `gamma.2sm[recipient.ID]`. 
+One then uses `2SM_Send` to encrypt the message and store the updated protocol in `gamma`.
+
+```
+if gamma.2sm[recipient_ID] = empty_string then
+gamma.2sm[recipient_ID] = 2SM_Init(gamma.myID, recipient_ID)
+(gamma.2sm[recipient_ID], ciphertext) = 2SM_Send(gamma.2sm[recipient_ID], plaintext)
+return (\gamma, ciphertext)
+```
+
+#### decrypt-from
+
+After receiving the sender’s `ID` and a ciphertext, 
+it behaves as the reverse function of `encrypt-to` and has a similar initialization:
+
+```
+if gamma.2sm[sender_ID] = empty_string then
+gamma.2sm[sender_ID] = 2SM_Init(gamma.myID, sender_ID)
+(gamma.2sm[sender_ID], plaintext) = 2SM_Receive(gamma.2sm[sender_ID], ciphertext)
+return (gamma, plaintext)
+```
+
+#### update-ratchet
+
+This function generates the next update secret `I` for the group member `ID`. 
+The ratchet state is stored in `gamma.ratchet[ID]`. 
+It is required to use a HMAC-based key derivation function HKDF to combine the ratchet state with an input, 
+returning an update secret and a new ratchet state.
+
+```
+(updateSecret, gamma.ratchet[ID]) = HKDF(gamma.ratchet[ID], input)
+return (gamma, updateSecret)
+```
+
+#### member-view
+
+This function calculates the set of group members based on the most recent control message sent by the specified user `ID`. 
+It filters the group membership operations to include only those observed by the specified `ID`, and 
+then invokes the DGM function to generate the group membership.
+
+```
+ops = {m in gamma.history st. m was sent or acknowledged by ID}
+return DGM(ops)
+```
+
+#### generate-seed
+
+This functions generates a random bit string and 
+sends it encrypted to each member of the group using the `2SM` mechanism. 
+It returns the updated protocol state and 
+the set of direct messages to send.
```

Comment:
```suggestion
the set of direct messages (denoted as `dmsgs`) to send.
```

Because I spend time to find difference between msg and dmsg in next functions description